components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java

```diff
@@ -1783,134 +1783,370 @@ private int handleStartIndexEqualsNULL(Integer startIndex) {
+            Set<String> roleNames = getRoleNamesForGroupsEndpoint(domainName);
             for (String roleName : roleNames) {
                 String userStoreDomainName = IdentityUtil.extractDomainFromName(roleName);
                 if (isInternalOrApplicationGroup(userStoreDomainName) || isSCIMEnabled(userStoreDomainName)) {
                     if (log.isDebugEnabled()) {
-                        log.debug("SCIM is enabled for the user-store domain : " + userStoreDomainName + ". " +
-                                "Including group with name : " + roleName + " in the response.");
+                        log.debug("SCIM is enabled for the user-store domain : " + userStoreDomainName + ". "
+                                + "Including group with name : " + roleName + " in the response.");
                     }
                     Group group = this.getGroupWithName(roleName);
                     if (group.getId() != null) {
                         groupList.add(group);
                     }
                 } else {
                     if (log.isDebugEnabled()) {
-                        log.debug("SCIM is disabled for the user-store domain : " + userStoreDomainName + ". Hence " +
-                                "group with name : " + roleName + " is excluded in the response.");
+                        log.debug("SCIM is disabled for the user-store domain : " + userStoreDomainName + ". Hence "
+                                + "group with name : " + roleName + " is excluded in the response.");
                     }
                 }
             }
         } catch (org.wso2.carbon.user.core.UserStoreException e) {
             String errMsg = "Error in obtaining role names from user store.";
             errMsg += e.getMessage();
             throw new CharonException(errMsg, e);
+        } catch (org.wso2.carbon.user.api.UserStoreException e) {
+            String errMsg = "Error in retrieving role names from user store.";
+            throw new CharonException(errMsg, e);
         } catch (IdentitySCIMException | BadRequestException e) {
             throw new CharonException("Error in retrieving SCIM Group information from database.", e);
         }
         //set the totalResults value in index 0
-        groupList.set(0, groupList.size()-1);
+        groupList.set(0, groupList.size() - 1);
         return groupList;
     }
 
+    private Set<String> getRoleNamesForGroupsEndpoint(String domainName)
+            throws UserStoreException, IdentitySCIMException {
+
+        SCIMGroupHandler groupHandler = new SCIMGroupHandler(carbonUM.getTenantId());
+        if (StringUtils.isEmpty(domainName)) {
+            return groupHandler.listSCIMRoles();
+        } else {
+            // If the domain is specified create a filter with SW operator and retrieve the role names by
+            // appending the domain name.
+            String attributeValue = domainName + CarbonConstants.DOMAIN_SEPARATOR;
```

Comment:
Shall we modify the name `attributeValue`, since this is used as an identifier?

Reply:
This block follows the same approach as a filter. The line actually creates and attribute value to filter the role names. So modifying the identifier is not needed since it conveys the actual usage of the identifier.